app/src/main/java/de/westnordost/streetcomplete/quests/bollard_type/AddBollardType.kt

```diff
@@ -0,0 +1,44 @@
+package de.westnordost.streetcomplete.quests.bollard_type
+
+import de.westnordost.streetcomplete.R
+import de.westnordost.streetcomplete.data.elementfilter.toElementFilterExpression
+import de.westnordost.streetcomplete.data.osm.edits.update_tags.StringMapChangesBuilder
+import de.westnordost.streetcomplete.data.osm.mapdata.Element
+import de.westnordost.streetcomplete.data.osm.mapdata.MapDataWithGeometry
+import de.westnordost.streetcomplete.data.osm.osmquests.OsmElementQuestType
+
+class AddBollardType : OsmElementQuestType<BollardType> {
+
+    private val bollardNodeFilter by lazy { """
+        nodes with
+            barrier = bollard
+            and !bollard
+    """.toElementFilterExpression() }
+
+    override val commitMessage = "Add bollard type"
+    override val wikiLink = "Key:bollard"
+    override val icon = R.drawable.ic_quest_no_cars
+    override val isDeleteElementEnabled = true
+
+    override fun getTitle(tags: Map<String, String>) = R.string.quest_bollard_type_title
+
+    // exclude free-floating nodes
+    override fun getApplicableElements(mapData: MapDataWithGeometry): Iterable<Element> {
+        val wayNodeIds = mutableSetOf<Long>()
+        mapData.ways
+            .filter { it.tags["area"] != "yes" }
+            .flatMapTo(wayNodeIds) { it.nodeIds }
```

Comment:
I wonder if for this here and other quests that work similarly, it would make a noticeable performance difference if this was used instead:

```suggestion
        mapData.ways.asSequence()
            .filter { it.tags["area"] != "yes" }
            .flatMapTo(wayNodeIds) { it.nodeIds }
```
Because the result of the `.filter` will be a list of almost all ways in the map data. I.e. almost the whole list of ways is copied. On the other hand, I don't know how big is the overhead of using `asSequence`

Care to do some performance testing?

Reply:
(and if it is a noticeable performance difference, maybe it would make sense if `ways` actually returned a sequence already? Or maybe even that `MapData` didn't implement `Iterable` but `Sequence`?)

Reply:
> Care to do some performance testing?

How would I do it?

Reply:
You could just write a test or a small kotlin program (with `fun main()`) and simulate the situation, measure the time with System.currentTimeMillis().

Reply:
This can also be convenient, sometimes: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.system/measure-time-millis.html

Reply:
I'm not sure how to simulate the situation in a test or script, because I need real map data there.

I did a small comparison of the processing times of the [OsmQuestController](https://github.com/streetcomplete/StreetComplete/blob/da5cb321546bace9ba39910bad6977051d01c6d1/app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/OsmQuestController.kt#L145-L156) though (exact same location, always already filled database):

* without `asSequence`:

  ```
  AddBollardType: Found 0 quests in 23ms
  AddBollardType: Found 0 quests in 15ms
  AddBollardType: Found 0 quests in 28ms
  AddBollardType: Found 0 quests in 20ms
  AddBollardType: Found 0 quests in 30ms
  ```

  average: 23.2ms

* with `asSequence`:
  ```
  AddBollardType: Found 0 quests in 53ms
  AddBollardType: Found 0 quests in 29ms
  AddBollardType: Found 0 quests in 23ms
  AddBollardType: Found 0 quests in 21ms
  AddBollardType: Found 0 quests in 17ms
  ```

  average: 28.6ms

Reply:
Interesting, so the sequence seems to have a considerable overhead and/or copying lists is really fast.